Geometry transformations

willcohen opened the pull request:
For #8, an implementation:

**Implemented**
- Add proj4j dependency
- Add awareness of SRIDs to `geo.jts` construction functions
- Add helper functions to determine the number of dimensions in a geom, to enable the creation of EWKBs that are SRID-aware
- Add `read-wkb-hex`and `to-wkb-hex` functions that can read/write hex string version of WKB
- Add `to-ewkb`, and `to-ewkb-hex` functions that can write binary and hex string objects with SRIDs
- Add transformation of JTS geoms and coords between SRIDs, CRS names, and proj4 strings
- Modification to Shapelike protocol to integrate transformations and conversions to JTS.
- Check that all functions accepting Shapelike objects are properly aware of projections when necessary. 
  - resegment has been changed to first transform its linestring to WGS84, and then transforms all the output linestrings back to the original projection.
- Check that all functions accepting Point objects are also aware of projections, since some JTS Points may be using different projections now
  - distance operates on geohash points, so those should already be projection-aware now
- Add tests

willcohen commented:
@worace After thinking about how to extend this to protocols a little more, I'm also thinking that `org.locationtech.jts.geom.Coordinate` might need to be excluded from the `Point` protocol. `WGS84Point`, JTS `Point` (when given a particular corresponding SRID, which is always WGS84 before this pull request), and spatial4j `Point` all refer to unambiguous locations on the earth. However, `Coordinate` is [purely cartesian](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/Coordinate.html).

Adding `(to-srid [srid])` to the `Point` protocol (and, additionally, `Shapelike`) should be straightforward for everything but coordinate, since we can just call something like `(transform-geom (to-jts [this]) srid)` and get back a projected JTS object. But for `Coordinate`, we'll never know what projection it should be coming from. Casting it to WGS84 seems like the wrong move, since that `Coordinate` might actually just be a cartesian representation of a non-WGS84 projection and would then be silently and incorrectly transforming an object back to the earth.

worace commented:
>  I'm also thinking that org.locationtech.jts.geom.Coordinate should be removed from the Point protocol.

OK after reading through this and doing a bit more background reading I do see your point and am inclined to agree at least conceptually. `geo.spatial` is intended to deal with geodesic stuff and things that can be dealt with in earth units like meters etc, so Cartesian coords don't necessarily fit cleanly in that model.

On the other hand part this library's intent is to provide an easy to use "Do What I Mean" interface for some of these operations. In that context some users just want to assume everything is an earth point in the most generic off-the-shelf projection, and in the case of a coordinate just dumping it into WGS84 does a decent job of this.

So, if we are looking at taking this out, I'd want to make sure we still include a decent escape hatch for the "I just have a lat/lon and don't really know or care about the projection...please figure it out for me" use-case.

There is also a bit of compatibility concern where there are places in the library where we are already using coordinates in this way so it would take a bit of effort to migrate these (but definitely doable if we decide it is worth while).

So, thinking through these options a bit more it seems like we either:

1. Remove `org.locationtech.jts.geom.Coordinate` from the `Point` protocol. This maintains the conceptual purity of this abstraction, and saves us from "lying" to users by accepting their cartesian data as if it were Geodesic WGS84/4326 data. It introduces a bit of headache around compatibility and maybe more user friction since people who might have previously not thought about projections at all now have to think about it.
2. Leave it as is and just assume a `Coordinate` in this context represents a lat/lon encoded in the most generic SRID there is. This makes the interface more complete by allowing us to accept another type of point representation, but does have the potential to screw things up if people try to do spatial ops on Coordinates representing non-4326 data. It also saves us from compatibility issues, both internally within the library code and for users who may have been depending on this behavior.

Reflecting on these a bit more I guess I'm not sure I see what's so bad about **2**, but I would be interested to hear your thoughts on the matter. Do you think there is a high chance of people getting tripped up because we mis-converted Coordinates for them?

As one example you can currently use `geo.spatial/distance` to measure distance between 2 Coordinates, and it will just convert it to WGS84:

```clojure
(geo.spatial/distance (org.locationtech.jts.geom.Coordinate. -70.0024 30.0019)
                      (org.locationtech.jts.geom.Coordinate. -70.0025 30.0020))
=> 14.696091061340983
```

And PostGIS will do similarly:

```
select ST_Distance(ST_Point(-70.0024, 30.0019)::geography, ST_Point(-70.0025, 30.0020)::geography);
 st_distance 
-------------
 14.69609116
(1 row)
```

It seems like this is probably the most expected behavior for the majority of use-cases, even though it's technically making an unverifiable assumption about the source of the provided coordinates.

willcohen commented:
Thanks for looking at this, @worace!

I absolutely agree that the library needs the ability to just default to lat/long when possible, and that users shouldn't have to think about projections when they don't need to. That said, I also think that functionality is exactly what JTS `Point` is equipped to deal with. That's how it's worked so far in `geo` for points, with a factory that always uses 4326 for points it generates. This means that when the objects get passed to other things using JTS, they know how to handle them and everything works well.

The example you give – using `geo.spatial/distance` on two coordinates – is an example of something that I don't think should actually be possible. The purpose of those coordinate objects isn't to be spatially referenced on the earth and then used for calculations, but instead to be internal objects that build up into things like points or polygons that are themselves spatially referenced (or, for example, to be transformed one by one within a polygon using a proj4j transformation). 

If the goal of the library is to be "do what I mean," then I'd suggest that it shouldn't allow the use of internal-ish things and pretend that they're something else. If a user has two point-like objects and wants to think about doing operations on them, let's make sure they use a point, and let's default it to lat/lng. So, calling `spatial/distance` on two points should continue to be very possible – if they use a point. 

The way I dealt with this in the pull request is to say that when calling `(point x y)`, continue to default to WGS84, and only put a different projection in with a 3-argument `(point x y srid)` or 2-argument `(point coord srid)` where the first one is an (unprojected) JTS coordinate. That `coord` argument as an input still shows the importance of having a function like `(coordinate x y)` available that makes a JTS coordinate, even if that object type isn't elevated to act the same as a spatially referenced point right off the bat via the protocol.

In the PostGIS example, remember that you're still actually having to specify the projection system with `::geography`. Removing the cast to geography would make it closer to the role of what `Coordinate` is for, with PostGIS reporting cartesian distance between the coordinates, in this case using a slightly-crazy unit of degrees.

Let me know what you think. Is there a reason that you think a user should want to be doing all these operations on a coordinate instead of a point?

willcohen commented:
For things like `spatial/distance` which take coordinates as arguments, if the worry is breakage for existing users, maybe we could still remove `Coordinate` from the protocol and then for each of those functions:
- Check if the function is being passed two coordinates.
- If so, warn the user that they should be using points instead, and manually cast them into points before continuing the operation.

Additionally, this would mean that they'd also become aware of different projections. If you passed them JTS points that had different projections, we could convert to lat/long before constructing the line string.

willcohen commented:
Sorry @worace – last comment. I keep rethinking this!

Another option to minimize breakage would be to keep it in the protocol, except deprecate it and put warnings there instead of in all the other functions, which could get verbose. This would allow it to keep working for users who expect that behavior, while still putting it on a path to be removed. 

That said, if we're talking about bumping it to 2.0.0 anyway because the API is changing already, it seems like a not-unreasonable time to remove a logical inconsistency.

worace commented:
> The example you give – using geo.spatial/distance on two coordinates – is an example of something that I don't think should actually be possible.

Yeah I agree with you that it is technically incorrect, but in practice when someone asks for that it's almost certain that they mean "treat this as a lat/lon", which is what we do.

 It wouldn't be a huge burden to require users to turn stuff into points first (this is likely what we'd end up doing), but would be a change to several parts of the library.

> Is there a reason that you think a user should want to be doing all these operations on a coordinate instead of a point?

Maybe not explicitly creating a Coord and doing distance stuff with it, but more likely getting Coordinate objects produced by other JTS operations (or even other functions within this library). I know I have run into this myself for example when trying to do things like [this](https://github.com/Factual/geo/blob/master/src/geo/spatial.clj#L371).

We can obviously move to a world where we require you to first call `(point coord srid)` to translate it (it sounds like this is what we may end up doing). But it would be a change to several parts of the library.

Partly it's a question of what's the goal for supporting projections -- is it sufficient to have more-or-less bolt-on support for taking existing points or shapes (assumed to be WGS84) and translating them into other projections? If so that seems fairly straightforward and what you have implemented seems to get most of the way there.

Or, is it necessary to integrate awareness of projections into the rest of the library, such that if I ask for the distance of one point in CRS 4326 and another in CRS 26911, it recognizes that and negotiates a common reference between them?

I'd have to study up more on the ins and outs of projection handling in Spatial4J and Geohash.java but this seems fairly involved to me. Is it sufficient to just translate everything back to WGS84 when doing spatial operations? Even if that works, it seems like it could be a performance penalty for what would likely be a relatively small portion of usage. If you have done experiments into implementing this kind of translation I'd be curious to know what you came up with. Maybe it's not as bad as I am imagining.

willcohen commented:
Thanks for the reply! Here’s some background:

**Motivation**
- Much of our current spatial analysis, though enabled via Clojure, is highly dependent on PostGIS. Most municipal or state-level GIS data comes in in some kind of state plane projection, and there’s definitely value in doing calculations in those projections — the area of a feature, in a ftUS state plane, gives square feet, which makes sense for a building or property. I don't know of any good clojure libraries that are aware of dealing with projected JTS objects.
- It would reduce the burden on a backend database and allow Java to take on more of the processing of these datasets if the objects don’t always have to be converted to lat-long with a round-trip to PostGIS before entering Clojure-world. JTS and proj4j do enable this in Java, but to make this simple in Clojure such functionality either needs to get added into geo or implemented in some new wrapper of a wrapper library.
- We have a fair number of cases where we want to analyze this projected data relative to geohashes (and spatial4j, to the degree that this library already compares between the two): which geohash does this building or set of buildings fit in? What geohashes at a given precision cover a city’s boundaries? What are that geohash’s boundaries and area, and can we get that shape reprojected back into the original state plane for comparison to the original data? Doing this requires some kind of analysis that relates these non-JTS objects to projected JTS objects.

**Adding to geo versus enabling elsewhere**
- This library is the best wrapper of geospatial libraries I know of already, and it’s very handy that it translates between the three libraries. We've been using an internal fork of it since 2016, but moved back to using your direct version after last year's updates (with the jts 1.15 update also re-applied internally up until now). So, above all else, thanks for creating it in the first place. We find a lot of value in it!
- The fact that it doesn’t actually handle the SRID of valid JTS objects means that a user with projected data still has to pre-solve that on their own somehow, which currently means wrapping this library with other logic. It’s only an issue with non-lat-long data, but lots of GIS info is that.
- Geotools is a Java library that enables reprojection of JTS, but there’s a lot of boilerplate to make it happen, it needs a lot of dependencies, and it’s LGPL. proj4j can create the transformation, but applying it to all of a Geometry’s coordinates still means dropping down to Java and JTS, which is basically what these proposed transform functions take care of.
- The only way I can think of to add this externally would be to create a full fork instead of a simple wrapper, because adding functions to the existing Point/Shapelike protocols that are projection-aware can’t be done with extend-protocol. If you guys think it’s cleaner to keep this library lat/long only and avoid the possible overhead, especially as it relates to your purposes, and not dig into this mess of GIS and GIS projections, that does make a lot of sense too. I just didn't want to pre-emptively fork a library we already find very valuable!

**Goals for a change:**
- All valid JTS geometries should work without having to think about what projection they're coming in as. If the projection isn't set, just call it lat-lng.
- Transforming a JTS geometry's projection should be as easy as converting it to a spatial4j or geohash object. This means adding some sort of `(to-jts)` with optional srid arguments into the actual protocols.
- Any implementation should minimize impact to objects that are already in lat/lng, and a good assumption for objects being created is that unless specified otherwise, the projection always defaults to lat/lng.

**Ancillary benefits:**
- Transforming any java geospatial object into a projected JTS object and vice-versa could be fairly seamless.

**Overall options:**
After playing with the protocols some more, I realize there’s actually possibly two potential paths here, at a high level.

1. One is to try to add something like `(to-jts [this] [this srid])` exclusively to the Shapelike protocol (in other words, ignoring #7). In some ways, this avoids some of the confusion discussed in previous comments. Assuming protocols can be inherited – which I need to read up on – this would sidestep the coordinate debate, since those aren’t actually subclasses of geometries. The catch to this approach is that this doesn’t on its own allow for the conversion of, say, a geohash WGS84Point to a projected JTS Point in one step, which is actually the very thing that led me down this path in the first place last week. We’d have to do something like `(to-jts (to-jts-point geohash-point) srid)`, which feels a little redundant. In trying this path out, I’m actually having a hard time getting something defined as a Point to be able to take arguments set up in the protocol for Geometry, so I’m not sure it’s actually possible, but it is worth considering.

2. A second option is to extend it to both Point and Shapelike, as I’ve been assuming so far. This would mean that you could run something like `(to-jts-point geohash-point srid)` or `(to-jts-geom shape-like srid)`, and in both cases get back a projected JTS object in one step, and excluding the srid would put it in WGS84. The catch here is that now we’d have to make a decision around Coordinates and how to handle them, and think about the functions that currently operate under the assumption that everything is in lat/lng.

No matter what, I do think that most of the projection handling can lie in the actual protocols, where they convert between different types. The main times the transformations should come into play is if the object is a JTS object in a non-standard projection, or a user wants to turn an object into one with a different projection. Here's a potential version of such a protocol for Point, also with an plan of maybe leaving Coordinate in but marking that it shouldn't be used this way:

```clojure
(defprotocol Point
  (latitude [this])
  (longitude [this])
  (to-spatial4j-point [this])
  (to-geohash-point [this])
  (to-jts-point [this] [this srid]))

(defn deprecate-coordinate
  [f]
  (log/warn "Deprecated: Assuming that this coordinate uses lat/lng values. Migrate to use a JTS Point instead.")
  f)

(extend-protocol Point
  WGS84Point
  (latitude [this] (.getLatitude this))
  (longitude [this] (.getLongitude this))
  (to-spatial4j-point [this] (spatial4j-point this))
  (to-geohash-point [this] this)
  (to-jts-point ([this] (jts-point this))
                ([this srid] (jts/transform-geom (jts-point this) srid)))

  org.locationtech.jts.geom.Point
  (latitude [this] (.getY (jts/transform-geom this 4326)))
  (longitude [this] (.getX (jts/transform-geom this 4326)))
  (to-spatial4j-point [this] (spatial4j-point this))
  (to-geohash-point [this] (geohash-point this))
  (to-jts-point ([this] this)
                ([this srid] (jts/transform-geom this srid)))

  org.locationtech.jts.geom.Coordinate
  (latitude [this] (deprecate-coordinate (.y this)))
  (longitude [this] (deprecate-coordinate (.x this)))
  (to-spatial4j-point [this] (deprecate-coordinate (spatial4j-point this)))
  (to-geohash-point [this] (deprecate-coordinate (geohash-point this)))
  (to-jts-point ([this] (deprecate-coordinate (jts-point this)))
                ([this srid] (deprecate-coordinate (jts/transform-geom (jts-point this) srid))))

  org.locationtech.spatial4j.shape.Point
  (latitude [this] (.getY this))
  (longitude [this] (.getX this))
  (to-spatial4j-point [this] this)
  (to-geohash-point [this] (geohash-point this))
  (to-jts-point ([this] (jts-point this))
                ([this srid] (jts/transform-geom (jts-point this) srid))))
```

and for Shapelike, something like:
```clojure
(defprotocol Shapelike
  (^Shape to-shape [this] "Convert anything to a Shape.")
  (^Geometry to-jts-geom [this] "Convert anything to a JTS Geometry."
                         [this srid] "Convert anything to a projected JTS Geometry."))

(extend-protocol Shapelike
  Shape
  (to-shape [this] this)
  (to-jts-geom ([this] (.getGeom this))
               ([this srid] (jts/transform-geom (.getGeom this) srid)))

  Geometry
  (to-shape [this]
    ;; Cloning geometries that cross dateline to workaround
    ;; spatial4j / jts conversion issue: https://github.com/locationtech/spatial4j/issues/150
    (let [geom (if (crosses-dateline? this)
                   (.clone this)
                   this)
          dateline-180-check? true
          allow-multi-overlap? true]
      (.makeShape jts-earth geom dateline-180-check? allow-multi-overlap?)))
  (to-jts-geom ([this] this)
               ([this srid] (jts/transform-geom this srid))))
```

For the functions that currently operate on JTS objects, some additional checking probably needs to happens: in `resegment`, there probably will need to be a step in there where the LineString is converted to WGS84 if need be. Its internal helpers like `under-cap-with-next-point?` can reasonably expect to only be called with things that continue to be WGS84 coordinates, so it's probably fine to ignore projections there. Things like `point-between` should be able to have coords switched out for Points (casting coords to points if needed to reduce breakage). Since it relies on `distance`, which itself relies on `to-geohash-point`, if the protocol is aware of projections, things should be okay.

I'm less sure about what to suggest for the functions in `poly.clj`. On the one hand, many of those operations seem projection-agnostic, so maybe its nice to have them work in coordinate systems. On the other hand, I don't think they're directly related to much happening in JTS. Perhaps they stay lat/lng focused for now and then additional wrappers around actual JTS functionality like union and intersection eventually get added in too? I suppose when actually testing all this out that might need to get revisited.

Let me know what you think!

willcohen commented:
Great!

For now, since everything is lat long here, I'm just using WKTs and transforming things into and out of 4326 in PostGIS. I think WKT is always projection-less, so on postgis it's something like `ST_SetSRID(ST_GeomFromText(wkt), 4326)`, and potentially wrapping that in another `ST_Transform()`. PostGIS has a SRID-aware extension of WKB called EWKB, which [JTS can write and read](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/io/WKBWriter.html).

I have another project deadline the next few days, but unless you're itching to get this 2.0 release out immediately, it'd be great for me to do a little homework with respect to all this. If you're okay with waiting on 2.0 until projections get figured out, I agree that it makes sense to think it through into full use cases like this.

I'm pretty sure PostGIS returns an EWKB or something close to it when asking for a geometry – try `SELECT geom, ST_AsEWKB(geom) FROM exampletable`, unless it's just making them appear similar for display for an end user – so as long as JDBC returns something similar, as long as that can eventually get into an object type that the JTS WKBReader can work with, I think it would be able to read them directly-ish. There'd probably be a little bit of change to `geo.io` to handle projections too, but nothing big.

Lastly, you're right that to-shape needs a few checks here and there, and that's a good catch. I forgot to add in the previous comment that the conversion functions would need a few modifications to handling their inputs as well, in addition to the actual protocol changes either way.

Thanks!

worace commented:
I guess they also have a WKT version which includes SRID which they call EWKT, but no idea if this is a widely supported format or not: http://www.postgis.net/docs/ST_AsEWKT.html

I'm not in any particular rush to get out the 2.0 release so I'm happy to come back to it next week if you like. If I can find some time I will do a bit of research into reading EWKB / EWKT from JTS as well

willcohen commented:
@worace I updated the PR to operate on Shapelike instead of Point. This totally sidesteps the whole Coordinate issue and seems to work well, since Geometry trickles down to JTS's Point. I also added WGS84Point to the extended Shapelike in `geohash.clj`, so I think that covers pretty much everything. The spatial4j `Circle`, I don't think, can be directly cast to a JTS geometry, so I think we only need to cover Rectangle and Point as seen in the PR.

I still haven't dug deeper into the EWKB issue but if this approach to transformations looks good, I can start thinking about that now as well.

willcohen commented:
@worace Looks like PostGIS could integrate quite nicely. I think some pretty small steps would be sufficient to integrate. The actual database back and forth should definitely live outside this library, but extending Shapelike to PGobject and then adding a corresponding `(to-pg-geom)` or something like it would mean that users could query a spatial table and act on what is returned or add/update rows directly, without having to wrap all geometry columns in conversion functions:

- Add `[org.postgresql/postgresql]` as a dependency, so that the library is aware of native objects.
- Add helper functions to geo.io that perform the read and write from/to PGobjects
- Extend Shapelike using PGobject and add a corresponding `(to-pg-geom)`, and perhaps `(to-pg-geog)`

### PGobject IO

Running a PostGIS query via JDBC that creates a geometry: `SELECT ST_SetSRID(ST_GeomFromText('POINT (10 10)')) AS geom`, returns an object like this:

```java
#object[org.postgresql.util.PGobject 0x7efa9ea1 "010100000000000000000024400000000000002440"]
```

We can actually test that these objects are postgis geometries before doing operations on them:

```clojure
(.getType obj-from-that-query)
=> "geometry"
```

They can also be converted to hex strings using postgres jdbc:

```clojure
(.toString obj-from-that-query)
=> "010100000000000000000024400000000000002440"
```

### Unprojected Point

The WKB reader can turn those hex strings to JTS:
```clojure
(.read wkb-reader (WKBReader/hexToBytes "010100000000000000000024400000000000002440"))
=> #object[org.locationtech.jts.geom.Point 0x515c6a62 "POINT (10 10)"]
```

A geometry without an SRID returns 0:
```clojure
(.getSRID (.read wkb-reader (WKBReader/hexToBytes "010100000000000000000024400000000000002440")))
=> 0
```

### Projected Point

With a projected geometry, `SELECT ST_SetSRID(ST_GeomFromText('POINT (10 10)'), 3586) AS geom`:

```java
#object[org.postgresql.util.PGobject 0x5c78b615 "0101000020020E000000000000000024400000000000002440"]
```

```clojure
(.read wkb-reader (WKBReader/hexToBytes "0101000020020E000000000000000024400000000000002440"))
=> #object[org.locationtech.jts.geom.Point 0x2dfb74a4 "POINT (10 10)"]
```
```clojure
(.getSRID (.read wkb-reader (WKBReader/hexToBytes "0101000020020E000000000000000024400000000000002440")))
=> 3586
```

### Unprojected Polygon

For an unprojected polygon, `SELECT ST_GeomFromText('POLYGON((0 0, 1 0, 1 1, 0 1, 0 0))') AS geom`:

```java
#object[org.postgresql.util.PGobject
        0x56d5bae3
        "0103000000010000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F0000000000000000000000000000F03F00000000000000000000000000000000"]
```
```clojure
(.read wkb-reader (WKBReader/hexToBytes "0103000000010000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F0000000000000000000000000000F03F00000000000000000000000000000000"))
=> #object[org.locationtech.jts.geom.Polygon 0x69044e73 "POLYGON ((0 0, 1 0, 1 1, 0 1, 0 0))"]
```

### Projected Polygon

And lastly, a projected polygon, `SELECT ST_SetSRID(ST_GeomFromText('POLYGON((0 0, 1 0, 1 1, 0 1, 0 0))'), 3586) AS geom`:

```java
#object[org.postgresql.util.PGobject
        0x3c811f80
        "0103000020020E0000010000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F0000000000000000000000000000F03F00000000000000000000000000000000"]
```

```clojure
(.read wkb-reader (WKBReader/hexToBytes "0103000020020E0000010000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F0000000000000000000000000000F03F00000000000000000000000000000000"))
=> #object[org.locationtech.jts.geom.Polygon 0x67092561 "POLYGON ((0 0, 1 0, 1 1, 0 1, 0 0))"]
```

```clojure
(.getSRID (.read wkb-reader (WKBReader/hexToBytes "0103000020020E0000010000000500000000000000000000000000000000000000000000000000F03F0000000000000000000000000000F03F000000000000F03F0000000000000000000000000000F03F00000000000000000000000000000000")))
=> 3586
```

willcohen commented:
I've gone through and added a bunch of tests, both to convert to and from JTS objects, as well as to check that both single points as well as geometries are able to be transformed between projection systems. 

I think this PR is looking good to merge. Let me know what else you’d like to see!

@worace

worace commented:
@willcohen I haven't been able to dig through too carefully but at first glance this looks pretty good! Thanks for all the work that went into it.

Unfortunately I am traveling for the next 10 days or so, so I may be a little slow to respond in more depth. But I think this looks like a good direction and I just want to take some time to read it more closely before merging.

As for the postgres stuff I'd be interested to discuss it further as a separate issue once we have this batch of changes resolved. My initial reaction is that I'm skeptical about adding any direct connections to postgresql or JDBC from this library as I think it's kind of beyond the scope of what we're providing. 

But it seems like once we provide the easy interface via EWKB that it should be pretty easy for users to integrate with their own DB code and maybe that solves 90% of the problem for most people anyway.

willcohen commented:
I closed #15 in favor of just adding some hex string IO here, which doesn't need any new dependencies. Like you suggest, this does get the conversion 90-ish% to being ready for postgis. I'll try to put this through its paces a little bit and try it out on a real database a little more fully to make sure it works out well and if there's anything that should happen here without expanding the library's scope too much. 

Once all this stuff lands, there might be value in thinking through some kind of 2.0-beta release to allow people to put this through its paces before a major version bump, given the scope of changes since 1.2. @worace

willcohen commented:
@worace are there other changes you'd like to see here?

worace commented on src/geo/io.clj:
```diff
@@ -21,8 +23,38 @@
+
+(defn to-ewkb [^Geometry geom]
+  "Write an EWKB, including the SRID"
+  (let [dims (jts/geom-dimension-check geom)]
```
I think JTS has a method for this: https://locationtech.github.io/jts/javadoc/org/locationtech/jts/geom/Geometry.html#getDimension()

Is it possible to use that here? In general I'd like to avoid anything that requires traversing all the coordinates of a geometry

willcohen replied:
Yes, I wasn't super pleased to end up having to pass through the coordinates like that. I was originally trying to use .getDimension but I think that's something different -- the javadoc states that "Returns the dimension of this geometry. The dimension of a geometry is is the topological dimension of its embedding in the 2-D Euclidean plane. In the JTS spatial model, dimension values are in the set {0,1,2}.

Note that this is a different concept to the dimension of the vertex Coordinates. The geometry dimension can never be greater than the coordinate dimension. For example, a 0-dimensional geometry (e.g. a Point) may have a coordinate dimension of 3 (X,Y,Z)."

I don't know of any function that can determine that "coordinate dimension", but I'll look a little more closely. It'd be nice to speed that up.

worace commented on src/geo/geohash.clj:
```diff
@@ -83,8 +86,8 @@
     ; counterclockwise.
-    (let [len   (* 4 (dec n))      ; Total sequence length
-          west  0                  ; We start (at L), with i = 0, going west
+    (let [;len   (* 4 (dec n))     ; Total sequence length
```
Let's remove these commented lines

willcohen replied:
There's a couple of places across the PR that have things commented out. I just didn't want to be too bold in deleting unused things entirely, but I'll apply this across the board.

worace commented:
@willcohen sorry I have been slow to get back to this; have been busy catching up after vacation. Will try to look at this more closely in the next couple days.

worace commented on test/geo/t_io.clj:
```diff
@@ -1,9 +1,14 @@
+(def wkb-hex "00000000030000000100000005C0518027525460AA403E007C84B5DCC6C0518027525460AA403E0068DB8BAC71C051801BDA5119CE403E0068DB8BAC71C051801BDA5119CE403E007C84B5DCC6C0518027525460AA403E007C84B5DCC6")
+(def ewkb-hex-wgs84 "0020000003000010E60000000100000005C0518027525460AA403E007C84B5DCC6C0518027525460AA403E0068DB8BAC71C051801BDA5119CE403E0068DB8BAC71C051801BDA5119CE403E007C84B5DCC6C0518027525460AA403E007C84B5DCC6")
+(def wkb-2-hex "00000000030000000100000019C05D9AA4FEC7483740410748DCF001DDC05D9AA54A2FCA834041074874966BEDC05D9AA7CF462C3340410744F7DB6700C05D9AA86ECD473E404107441B4FD3A0C05D9AC2D7FC35CB4041071F8E14EC8DC05D9AC2F86FEAC44041071FD93BC96BC05D9ACBCFE5D7EB404107345BC3A32BC05D9AC753BE8EA24041073A871AC7EEC05D9AC75D3578344041073A9A66DD10C05D9AC6B6EB509F4041073B7F1113F6C05D9AC6AD5119E44041073B6C003FB1C05D9 ...
+(def ewkb-2-hex-wgs84 "0020000003000010E60000000100000019C05D9AA4FEC7483740410748DCF001DDC05D9AA54A2FCA834041074874966BEDC05D9AA7CF462C3340410744F7DB6700C05D9AA86ECD473E404107441B4FD3A0C05D9AC2D7FC35CB4041071F8E14EC8DC05D9AC2F86FEAC44041071FD93BC96BC05D9ACBCFE5D7EB404107345BC3A32BC05D9AC753BE8EA24041073A871AC7EEC05D9AC75D3578344041073A9A66DD10C05D9AC6B6EB509F4041073B7F1113F6C05D9AC6AD5119E4404107 ...
```
@willcohen I realize you didn't start the trend of putting these examples inline but now that we're adding so many more of them I'd like to move them to separate files. Ideally we can put each one in its own file under `test/resources` and then just slurp those files into these definitions.

willcohen replied:
Will do!

willcohen commented:
No problem, I hope you had a good vacation! Just wanted to keep this moving.

worace commented on src/geo/jts.clj:
```diff
@@ -1,37 +1,138 @@
 (ns geo.jts
   "Wrapper for the locationtech JTS spatial library. Constructors for points,
   coordinate sequences, rings, polygons, multipolygons, and so on."
+  (:require [clojure.string :as str])
```
Small issue but I'd rather not use this alias since it shadows the default `str` function -- if anything I'd just refer the ones you're using like `(:require [clojure.string :refer [join]])` or whatever

worace commented on src/geo/jts.clj:
```diff
@@ -1,37 +1,138 @@
+                                      GeometryFactory)
+           (org.osgeo.proj4j CoordinateTransformFactory
+                             CRSFactory
+                             ProjCoordinate)))
```
@willcohen I think there's enough functionality related to projections here to warrant its own namespace; how do you feel about moving all of the new proj functions here into something like `geo.proj`?

willcohen replied:
That's an interesting idea and I'll have to think about that over the weekend. I suspect, however, that it might be impossible to do cleanly because of circular dependencies. Transforming a coordinate requires making a new coordinate, which itself uses the JTS namespace. It may seem silly to make a new namespace if the primary transformation functions are still living in JTS.

willcohen replied:
9dabeda would work! My one concern (and it might not be relevant) is that create-transform will return a proj4j transformation object. proj4j is [currently incubating at locationtech](https://github.com/locationtech/proj4j) (incidentally, this is why I'm currently including the geotrellis version of proj4j. The [original proj4j](https://github.com/Proj4J/proj4j) is getting quite long in the tooth, and the geotrellis version is the most recent version that's available on Maven Central. It's what the locationtech version is based on. That locationtech version [will have a different namespace](https://github.com/locationtech/proj4j/blob/master/src/main/java/org/locationtech/proj4j/CRSFactory.java#L18), like JTS, and I was just trying to hide those objects to avoid an eventual 3.0 version bump due solely to proj4j. It definitely seems strange to me to have the important create-transform be public and live in JTS but the little helpers in proj, which is why I kept it all in the same place.

If, on balance, splitting the namespace and preparing for an eventual version bump seems cleaner than exposing an internal transformation, I'll definitely switch the functions around.

willcohen commented:
I made changes in the latest commit, addressing everything but the proj namespace. Since JTS objects do have a place for SRIDs, some information about projections should still want to live there, like `get-srid` and `set-srid`. Similarly, I think the only way to remove the circular dependency may be to change `transform-coord` to drop down to java to make a coordinate, instead of using the `(coordinate)` function in `geo.jts`. While we could do that, it does seem like the library should be consuming its own functions whenever possible instead of duplicating functionality. We can still think about it some more, but I'm having a hard time drawing a consistent line of what would be jts and what would be proj.

Additionally, I can't find any built-in functions for checking whether or not a coordinate has a Z value, only what dimensions the geometry itself is -- a point in 3 dimensions and a point in 2 dimensions both have a ".getDimension" of 0. The only reason why this even matters is that the WKBWriter that is SRID-aware [requires that we specify how many dimensions the geometry has](https://locationtech.github.io/jts/javadoc/org/locationtech/jts/io/WKBWriter.html#WKBWriter(int,%20boolean)), and I think that may always require just looking at Z values for each coordinate in the geometry for now. 

Lastly, in terms of additional functionality cluttering the namespace, one mitigating factor may just be making as much private as possible. `(transform-geom)` is the most important function, and I think the helper functions I left public are still useful in general. As long as the proj4j interoperability stays private (making and applying the transforms, mostly), a lot of the new functionality can stay hidden-ish inside the library. More importantly, if proj4j changes its namespace once it's done incubating with locationtech, there won't be a need for a major version change again. I like the idea that for the purposes of this library, the only way "in" to working with projections is to provide SRIDs, projection names, or proj4 strings, rather than exposing all of the projection machinery and transformation objects.

@worace

worace commented on src/geo/jts.clj:
```diff
@@ -40,71 +143,206 @@
+    (isGeometryChanged [_]
+      true)))
+
+(defn- tf-set-srid
```
Again, my ignorance about these types of conversions and CRS's in general, but can you explain a bit more about how these 2 cases (SRID vs EPSG) are different, and whether it's really necessary to handle both?

I'd love to see some more explicit test cases that both verify these behavior and serve as documentation to explain how they might appear in normal usage.

willcohen replied:
JTS geometries only take integers as their projection system. If .getSRID is 4326, that refers to the EPSG:4326 coordinate system.
proj4j (see [comment](https://github.com/Factual/geo/pull/10/files/0dee51ffd1fd263ed7794ae87afd758b2b82aeff#r193432759)) takes a wider view.

The thought process here is:
1. When the coordinate system (either source or target) is provided as an integer, assume it's an EPSG, convert it into a string called "EPSG:xxxx" and pass that along.
2. When the coordinate system (either source or target) is provided as a string, first try to parse it as a coordinate system name (whether EPSG or any of the other authority options), and failing that then try to parse it as a proj4 string.
3. After the transformation occurs, tf-set-srid comes into play: if the target coordinate system represents an EPSG projection -- either as "EPSG:4326" or "4326", then JTS is able to store that in its srid metadata. If that's the case, do so during a projection.
4. When a target coordinate system represents anything else -- whether a string name that isn't "EPSG:" or proj4, then the transformation stills occur to the coordinates themselves, but nothing should be set for the JTS srid.
5. When the source and target projection systems are identical, clone the object without passing it through a transform. When that projection is an EPSG code, set the JTS's SRID to that.

Separately, there is an extraneous -> macro in transform-geom I also need to remove.

worace commented on src/geo/jts.clj:
```diff
@@ -40,71 +143,206 @@
+  ([wkt srid]
+   (multi-polygon (map #(polygon-wkt % srid) wkt))))
+
+(defn- coord-dimension-check
```
I'd love to talk a little more about the various 3-D checks in these functions. I'm curious if this is something that's required in order to be compatible/complete with proj4j or if we just think we should have it for completeness of the API.

These changes are already getting quite large and checking for 2d/3d support everywhere adds a lot of complexity. This library has never purported to support 3d operations before, so unless we are trapped into this by one of the dependencies I am starting to think I'd be happy to remove all the 3d stuff, at least for now. That would simplify this batch of changes quite a bit, and if we think it's really a big improvement we could always consider adding it as a separate round of changes.

willcohen replied:
I also have never needed 3d in my datasets. I don't think that this library needs to go out of its way to enable the creation of 3d shapes. If we drop all 3d support and stick to 2d for now, it wouldn't hurt my current use cases at all.

However, jts, ewkb, and geojson very much do support 3d. Proj4j isn't very good at 3d transformations, but if it gets a geometry with a lat, long, and 0 z dimension, it'll transform into a geometry with an x, y, and 0 z dimension. (In fact, projections are about to start thinking much more in earnest [about four dimensions](https://gdalbarn.com/), to account for changes in 3 dimensions over time!)

If we were using the full parent proj.4 instead of its pure java port, the 3d functionality in projections would be much more seamless. There is a JNI wrapper around proj.4, but using that would require that this library start dealing with keeping compiled versions of proj.4 for different platforms and integrating that, so I do think that sticking to proj4j is the right move. As proj4j starts to better handle 3d coordinates, I think that the loss of a third dimension becomes more noticeable.

Since the library can read formats in and its dependencies will create things with 3 dimensions, and since the dependencies can take 3 dimensional things and write them out correctly to the various formats, the one place where the library's core functionality may want to be 3d aware is where it interacts with coordinates and transformations. If we don't at least pass the z through the coordinate, any transformation will drop the z dimension on its way through the library. If 3d is removed from the library, we'll need to clarify that as a caveat in the io namespace.

worace commented on src/geo/spatial.clj:
```diff
@@ -94,23 +92,36 @@
   (>= (.getWidth (.getEnvelopeInternal jts-geom))
       180))
 
+(defn shape->geom
```
I'm wondering if this can be handled as part of the protocol, since that's really the ideal mechanism for doing type-based dispatch like this. It might need to be added as new cases in the `Shapelike` protocol, but it seems like these are really implementations of the `to-jts` portion of that protocol for the `RectangleImpl`, `JtsGeometry`, `PointImpl`, and `JtsPoint` classes

willcohen replied:
I don't know why I didn't think of this. This will mean removing `(to-jts)` from Shape, and only adding it to the types currently listed in `shape->geom` (excluding circles, basically), but it should definitely be faster.

worace commented on src/geo/spatial.clj:
```diff
@@ -293,12 +311,12 @@
-      square-degrees->steradians
-      steradians->area)))
+  ;(let [a (area-in-square-degrees rect)
+  ;      r (earth-radius (center rect))
```
Let's remove this comment.

willcohen commented on src/geo/jts.clj:
```diff
@@ -1,37 +1,140 @@
+              (integer? y-or-srid))
+         (.createPoint (gf y-or-srid) x-or-coordinate)))
+  ([x y srid]
+   (point (coordinate x y) srid)))
```
Between these two, 1 seems better to me too. I don't think it's that important to be making projected points from already-created-coordinates that often.

willcohen commented on src/geo/jts.clj:
```diff
@@ -40,71 +143,206 @@
+  "Check if two Geometries have the same SRID."
+  [^Geometry g1 ^Geometry g2]
+  (and (= (get-srid g1) (get-srid g2))
+       (not= (get-srid g1) 0)))
```
On the one hand, SRID 0 really just means "no projection", so it should be fine to have `same-geom?` report true for two geometries that have SRID 0.

On the other hand, if two geometries have, effectively, no SRID, it seems strange to me for `same-srid?` to report true.

Removing the 0 check will make `same-geom?` more sensible, and as long as this edge condition is noted in the doc string for `same-srid?`, I think removing the check should work.

willcohen commented on .gitignore:
```diff
@@ -10,3 +10,5 @@ build
 .lein-failures
 target/**
+.idea
+*.iml
```
Editor configuration it is -- sorry this kept coming up.

willcohen commented on src/geo/jts.clj:
```diff
@@ -1,37 +1,140 @@
+           (= (subs c 0 6) "WORLD:")
+           (= (subs c 0 6) "NAD27:"))
+       (catch Exception _
+         false)))
```
These come from [proj4j](https://github.com/Proj4J/proj4j/blob/master/trunk/src/main/java/org/osgeo/proj4j/CRSFactory.java#L46) and represent the range of authorities that proj4j recognizes for valid CRS names.

willcohen commented on src/geo/geohash.clj:
```diff
@@ -29,11 +23,20 @@
+                     (.getMinLon box)
+                     (.getMaxLon box)
+                     (.getMinLat box)
+                     (.getMaxLat box))))
```
This does seem simpler!

willcohen commented:
This pulls in the proj4j-review branch changes, removing 3d etc. I fixed up the tests from that -- the main issue was that I think `point` needs to have arities `[coordinate]`, `[lat long]` (instead of `[x y]`), and `[x y srid]` to be consistent with the other functions and for the protocols to work cleanly. Besides that, I added tests about the CRS names and exception throwing, removed the 0 check for `same-srid?`, and noted that in the doc strings for that and `same-geom?`.

Based on the Monday comments, I think the only thing unaddressed is the idea about the potential `srid?` function based on what's in transform-geom? It's not functionality that's appearing regularly so I think it's okay to leave transform-geom the way it is for now. @worace

willcohen commented:
Realizing I missed a final line ending in gitignore, just pushed that in as well today. Do you have additional questions about the handling around SRIDs/EPSGs/proj4 names/proj4 strings?

@worace

worace commented:
@willcohen no, i think it's looking pretty good; Thanks for working on that last round of updates. Sorry I have been silent on this but I think I will merge it in the next few days. Then I'll probably work on getting a couple more things merged and doing a bit of refactoring and extra testing in preparation for a 2.0 alpha release of some sort.

willcohen commented:
Great! Let me know if you need anything else on this one.
